model-engine/model_engine_server/inference/forwarding/http_forwarder.py

```diff
@@ -111,6 +118,47 @@ async def event_generator():
         return EventSourceResponse(event_generator())
 
 
+# This route is a catch-all for any requests that don't match the /predict or /stream routes
+# It will treat the request as a streaming request if the "stream" body parameter is set to true
+# NOTE: it is important for this to be defined AFTER the /predict and /stream endpoints
+# because FastAPI will match the first route that matches the request path
+async def predict_or_stream(
+    request: EndpointPredictV1Request,
+    background_tasks: BackgroundTasks,
+    sync_forwarder: Optional[Forwarder],
+    stream_forwarder: Optional[StreamingForwarder],
+    limiter=Depends(get_concurrency_limiter),
+):
+    if stream_forwarder and request.args and request.args.root.get("stream", False):
+        return stream(request, stream_forwarder, limiter)
+    elif sync_forwarder:
+        return predict(request, background_tasks, sync_forwarder, limiter)
+    else:
+        raise Exception("No forwarder configured for this route")
+
+
+def add_extra_routes():
+    """Read extra_routes from config and dynamically add routes to app"""
+    config = get_config()
+    # aggregate a list of routes -> (sync, stream) support
+    extra_routes: Dict[str, Tuple[Optional[Forwarder], Optional[StreamingForwarder]]] = dict()
+    for route in config.get("sync", {}).get("extra_routes", []):
+        extra_routes[route] = (load_forwarder(route), None)
+    for route in config.get("stream", {}).get("extra_routes", []):
+        extra_routes[route] = (
+            extra_routes.get(route, (None, None))[0],
```

Comment:
is this so that some routes are marked as both sync and streaming enabled?

Reply:
yeah.. e.g. if we use vllm openai server out of box, they use the same routes for sync and stream and use body:stream = true to distinguish requests